Refactor internal switch cases

manuelcandales opened the pull request:
There were multiple ET_INTERNAL_SWITCH_CASE_*_TYPES macros that repeated lists of types, for example:
`ET_INTERNAL_SWITCH_CASE_INT_TYPES`
`ET_INTERNAL_SWITCH_CASE_REAL_TYPES`
`ET_INTERNAL_SWITCH_CASE_FLOAT_TYPES`
`ET_INTERNAL_SWITCH_CASE_ALL_TYPES`
In this PR, we refactor them building from the bottom up, for example reusing INT and FLOAT to build REAL and reusing REAL to build ALL

Differential Revision: D72248082

swolchok commented on runtime/core/exec_aten/util/scalar_type_util.h:
```diff
@@ -921,55 +921,11 @@ struct promote_types {
-      ::executorch::aten::ScalarType::Bits8, CTYPE_ALIAS, __VA_ARGS__)         \
-  ET_INTERNAL_SWITCH_CASE(                                                     \
-      ::executorch::aten::ScalarType::Bits16, CTYPE_ALIAS, __VA_ARGS__)
+#define ET_INTERNAL_SWITCH_CASE_BOOL_TYPES(CTYPE_ALIAS, ...) \
+  ET_INTERNAL_SWITCH_CASE(                                   \
+      ::executorch::aten::ScalarType::Bool, CTYPE_ALIAS, __VA_ARGS__)
```
why does this macro exist? what other types could it contain that aren't exactly `ScalarType::Bool`?

swolchok commented on runtime/core/exec_aten/util/scalar_type_util.h:
```diff
@@ -979,12 +935,97 @@ struct promote_types {
+  ET_INTERNAL_SWITCH_CASE_ALLINT_TYPES(CTYPE_ALIAS, __VA_ARGS__) \
+  ET_INTERNAL_SWITCH_CASE_BOOL_TYPES(CTYPE_ALIAS, __VA_ARGS__)
+
+#define ET_INTERNAL_SWITCH_CASE_FLOAT_TYPES(CTYPE_ALIAS, ...)          \
   ET_INTERNAL_SWITCH_CASE(                                             \
       ::executorch::aten::ScalarType::Float, CTYPE_ALIAS, __VA_ARGS__) \
   ET_INTERNAL_SWITCH_CASE(                                             \
       ::executorch::aten::ScalarType::Double, CTYPE_ALIAS, __VA_ARGS__)
 
+#define ET_INTERNAL_SWITCH_CASE_ALLFLOAT_TYPES(CTYPE_ALIAS, ...)      \
```
having these both exist is confusing -- "Half and BFloat16 are part of all float types, but they're not float types"?

manuelcandales replied:
Half and BFloat16 are not part of the FLOAT set today. That's why we have FLOATH and FLOATHBF16.
I am using the prefix "ALL" to denote the set containing all of the types belonging to a given category (INT, FLOAT, REAL, COMPLEX, NUMERIC)

swolchok commented on runtime/core/exec_aten/util/scalar_type_util.h:
```diff
@@ -979,12 +935,97 @@ struct promote_types {
+  ET_INTERNAL_SWITCH_CASE(                                            \
+      ::executorch::aten::ScalarType::BFloat16, CTYPE_ALIAS, __VA_ARGS__)
+
+#define ET_INTERNAL_SWITCH_CASE_REAL_TYPES(CTYPE_ALIAS, ...)  \
+  ET_INTERNAL_SWITCH_CASE_INT_TYPES(CTYPE_ALIAS, __VA_ARGS__) \
+  ET_INTERNAL_SWITCH_CASE_FLOAT_TYPES(CTYPE_ALIAS, __VA_ARGS__)
+
+#define ET_INTERNAL_SWITCH_CASE_ALLREAL_TYPES(CTYPE_ALIAS, ...)  \
```
same -- if ET_INTERNAL_SWITCH_CASE_REAL_TYPES should not be deprecated, then we need better names that make it clear when you would use each one

swolchok commented on runtime/core/exec_aten/util/scalar_type_util.h:
```diff
@@ -979,12 +935,97 @@ struct promote_types {
+  ET_INTERNAL_SWITCH_CASE_ALLREAL_TYPES(CTYPE_ALIAS, __VA_ARGS__) \
+  ET_INTERNAL_SWITCH_CASE_BOOL_TYPES(CTYPE_ALIAS, __VA_ARGS__)
+
+#define ET_INTERNAL_SWITCH_CASE_COMPLEX_TYPES(CTYPE_ALIAS, ...)               \
+  ET_INTERNAL_SWITCH_CASE(                                                    \
+      ::executorch::aten::ScalarType::ComplexFloat, CTYPE_ALIAS, __VA_ARGS__) \
+  ET_INTERNAL_SWITCH_CASE(                                                    \
+      ::executorch::aten::ScalarType::ComplexDouble, CTYPE_ALIAS, __VA_ARGS__)
+
+#define ET_INTERNAL_SWITCH_CASE_ALLCOMPLEX_TYPES(CTYPE_ALIAS, ...) \
```
same concern as before -- confusing, why isn't the old one deprecated, etc.

swolchok commented on runtime/core/exec_aten/util/scalar_type_util.h:
```diff
@@ -1204,26 +1201,17 @@ struct promote_types {
-                  __VA_ARGS__)                                \
-                  ET_INTERNAL_SWITCH_CASE(                    \
-                      ::executorch::aten::ScalarType::UInt64, \
-                      CTYPE_ALIAS,                            \
-                      __VA_ARGS__))
+  ET_SWITCH_ALLREALB_TYPES(TYPE, CONTEXT, NAME, CTYPE_ALIAS, __VA_ARGS__)
```
this is more easily accomplished as `#define ET_SWITCH_REALHBBF16_AND_UINT_TYPES ET_SWITCH_ALLREALB_TYPES` (and similarly below)

swolchok reviewed (changes requested):
an up-to-date English explanation of the taxonomy of types in a comment is probably in order so people don't have to read through all the macros.

would also like @mergennachin's opinion on this

manuelcandales commented on runtime/core/exec_aten/util/scalar_type_util.h:
```diff
@@ -979,12 +935,97 @@ struct promote_types {
+  ET_INTERNAL_SWITCH_CASE(                                              \
+      ::executorch::aten::ScalarType::UInt64, CTYPE_ALIAS, __VA_ARGS__)
+
+#define ET_INTERNAL_SWITCH_CASE_ALLINT_TYPES(CTYPE_ALIAS, ...) \
```
ALLINT is the name of the set, just like REALH and REALHBF16 are names

swolchok reviewed (approved):
seems fine

manuelcandales committed:
Refactor internal switch cases (#9802)

Summary:
There were multiple ET_INTERNAL_SWITCH_CASE_*_TYPES macros that repeated lists of types, for example:
`ET_INTERNAL_SWITCH_CASE_INT_TYPES`
`ET_INTERNAL_SWITCH_CASE_REAL_TYPES`
`ET_INTERNAL_SWITCH_CASE_FLOAT_TYPES`
`ET_INTERNAL_SWITCH_CASE_ALL_TYPES`
In this PR, we refactor them building from the bottom up, for example reusing INT and FLOAT to build REAL and reusing REAL to build ALL


Reviewed By: swolchok

Differential Revision: D72248082
